Resolve type assertion

TripleDogDare opened the pull request:
This gives an error instead of a panic for the demo code base in #2. (#5 go1.18+ fix seems to have resolved the panic as well)

Now it returns a bunch of errors that seem much more reasonable, if not any more functional. There's a good chance that the codes from the old identity could be assigned to the new identity and just ignore the type assertion but I'm not really sure how to do that.
```
$ go-serum-analyzer -strict ./...
/home/cjb/repos/misc/go-demo-app-with-serum/cli/cli.go:107:1: function "Do" is exported, but does not declare any error codes
/home/cjb/repos/misc/go-demo-app-with-serum/jobbers/terror/terror.go:23:6: unsupported: tracking error codes across type assertions
/home/cjb/repos/misc/go-demo-app-with-serum/jobbers/terror/terror.go:21:1: function "InducePanic" has a mismatch of declared and actual error codes: unused codes: [jobber-error-natch]
```

TripleDogDare commented:
I think I got it to do the right thing!
I still have no idea what the `ast.TypeAssertExpr.X` refers to so it's just a guess. However, it _appears_ to do the right thing against the demo code branch.

TripleDogDare committed:
type assertion tests

TripleDogDare committed:
type assertion detection

TripleDogDare commented:
So this is less required now that it no longer panics. However, adding tests and better error messages for handling type assertion seems nice to have.

warpfork commented on analysis/taint_spread.go:
```diff
@@ -10,7 +10,7 @@ import (
 	taintSpreadResult struct {
 		expressions        []ast.Expr             // expressions that represent the taint, or nil
-		destructAssignment []*taintSpreadDestruct // taint originating from destructirung assignments, or nil
+		destructAssignment []*taintSpreadDestruct // taint originating from destructing assignments, or nil
```
```suggestion
		destructAssignment []*taintSpreadDestruct // taint originating from destructuring assignments, or nil
```

TripleDogDare committed:
update destructAssignment comment s/destructing/destructuring/

Co-authored-by: Eric Myhre <hash@exultant.us>

warpfork commented on analysis/analyse.go:
```diff
@@ -779,25 +788,25 @@ func findErrorCodesFromIdentTaint(c *context, visitedIdents map[*ast.Object]stru
+			newCodes := findErrorCodesInCallExpression(c, expr, function)
+			result = Union(result, newCodes)
+		default:
+			panic(fmt.Sprintf("destruct assignment should only originate from a call expression: type %T", expr))
```
```suggestion
			panic(fmt.Sprintf("destructuring assignment is currently only supported from a call expression: found one in an expression of type %T", expr))
```

I started with a typo fix...

... but also suggest a different wording, because I'm pretty sure there's more origins of these, fwiw.  `x, y := w, z` expressions exist and I think they're filed under destructuring assignments.  (Pushing that further support doesn't necessarily have to be part of this PR though.)

TripleDogDare replied:
That seems fair. I'll open an issue for it so we can track that.

warpfork reviewed (approved):
I think I see some more branches that should be implemented some day.  (Gosh, I wish golang had sum types, so we could simply have the compiler tell us when we're successfully exhaustively matching on things!!)

Looks like this is incrementally better, though.  More things covered, less panicking.  And it's got tests.  LGTM.

TripleDogDare committed:
update non-call destructuring assignment panic message

There is a possibility that destructuring assignment occurs via non-call expressions.

Co-authored-by: Eric Myhre <hash@exultant.us>
